### PR TITLE
virion 3.1

### PR DESCRIPTION
**Pull request by SOF3**

### CODE REVIEW

File: virion.md [hunk 1]
```diff
@@ -51,9 +58,62 @@ and SHOULD avoid using it as a substring in their fully-qualified paths
...
+  - Public static fields are allowed.
+  - Constant definitions must be resolvable at compile time.
+
+Items in the shared namespace MUST NOT reference any shaded items,
+including shaded items declared by transitive dependencies.
+The term "reference" here includes parameter types, return types,
+`extends`/`implements` target and attributes.
+
+Once a virion has been published,
+the defition of all items under the shared namespace root MUST NOT ever be changed.
```
Comment (dktapps) [inline virion.md:97]:
```suggestion
the definition of all items under the shared namespace root MUST NOT ever be changed.
```
this also seems a bit extreme

File: virion.md [hunk 2]
```diff
@@ -51,9 +58,62 @@ and SHOULD avoid using it as a substring in their fully-qualified paths
...
+In particular:
+
+- Semantics-sensitive attributes and doc comments MUST NOT be changed.
```
Comment (dktapps) [inline virion.md:100]: even doc comment mistakes can't be fixed?

**Comment by dktapps** [comment]
I get the purpose of this, but it feels way too limiting

File: virion.md (hunk 1 above)
Comment (JaxkDev) [reply virion.md:97]: +1

Maybe only changeable on major (from major.minor.patch)?

File: virion.md (hunk 2 above)
Comment (SOF3) [reply virion.md:100]: @dktapps "semantics-sensitive" refers to doc comments too. Such as `@noHandle` or something like that.

File: virion.md (hunk 1 above)
Comment (SOF3) [reply virion.md:97]: @dktapps There is no version control, so we must ensure that they are always mutually compatible.

Comment (dktapps) [reply virion.md]: I get that, but I still don't like it.

File: virion.md (hunk 2 above)
Comment (SOF3) [reply virion.md:100]: Normal doc comment changes are fine since they don't affect runtime.

Comment (dktapps) [reply virion.md:100]: @noHandle is defined by PM. What happens if PM changes this in the future?

Comment (SOF3) [reply virion.md:100]: ok maybe it is not so strictly required if two versions can't run simultaneously due to incompatible major. but better have stronger requirement to start with.

**Review by jasonw4331 — APPROVED** [review]

**Comment by SOF3** [comment]
Question: should we allow empty constructors with only promoted properties too?

**Comment by dktapps** [comment]
what do promoted properties have to do with anything?

**Comment by SOF3** [comment]
the current proposal disallows implementing functions in the shared namespace to avoid conflicting implementations, but it seems we could allow empty functions.

**Comment by dktapps** [comment]
I wasn't aware it disallows implementing functions, but even that doesn't make sense to me

**Comment by SOF3** [comment]
We consider the shared namespace to be a protocol interface instead of an actual util. In that sense, an immutable shared namespace is basically like how you are not allowed to change the fields of a protocol instead of adding new packet types etc.

File: virion.md [hunk 3]
```diff
@@ -36,12 +36,19 @@ It MUST also declare a `virion` extra attribute in the composer.json:
   "extra": {
     "virion": {
       "spec": "3.0",
```
[the remark this answers is not in the document]
Comment (SOF3) [reply virion.md:38]: Actually bumping the version is not strictly necessary if shared namespace is not used.

**Pull request merged**